src/Core/Vault/Services/Implementations/CipherService.cs

```diff
@@ -966,4 +987,35 @@ private async Task ValidateCipherCanBeShared(
 
         ValidateCipherLastKnownRevisionDateAsync(cipher, lastKnownRevisionDate);
     }
+
+    // This method is used to filter ciphers based on the user's permissions to delete them.
+    // It supports both the old and new logic depending on the feature flag.
+    private async Task<List<T>> FilterCiphersByDeletePermission<T>(
+        IEnumerable<T> ciphers,
+        HashSet<Guid> cipherIdsSet,
+        Guid userId) where T : CipherDetails
+    {
+        if (!_featureService.IsEnabled(FeatureFlagKeys.LimitItemDeletion))
+        {
+            return ciphers.Where(c => cipherIdsSet.Contains(c.Id) && c.Edit).ToList();
+        }
+
+        var filteredCiphers = new List<T>();
+        var user = await _userService.GetUserByIdAsync(userId);
+        var ciphersByOrg = ciphers.Where(c => cipherIdsSet.Contains(c.Id))
+            .GroupBy(c => c.OrganizationId);
+        var organizationAbilities = await _applicationCacheService.GetOrganizationAbilitiesAsync();
+
+        foreach (var group in ciphersByOrg)
+        {
+            var organizationAbility = group.Key.HasValue &&
+                organizationAbilities.TryGetValue(group.Key.Value, out var ability) ?
+                ability : null;
+
+            filteredCiphers.AddRange(group
+                .Where(c => NormalCipherPermissions.CanDelete(user, c, organizationAbility)));
+        }
+
+        return filteredCiphers;
```

Comment:
🤔 I think Linq's `SelectMany` can help flatten each group; I don't know if it makes any significant difference. This is just a thought, no significant changes are required